### PR TITLE
Add MFA support for Auth Emulator

### PR DESCRIPTION
**Pull request by yuchenshi**
### Description

This adds partial MFA support for password sign-in etc. Error messages are NOT yet done (strings in double braces may not match production) and signInWithIdp + MFA doesn't work right now.

Many tests for edge cases (especially error strings) TBD.

### Scenarios Tested

See tests added

### CODE REVIEW

File: src/emulator/auth/operations.ts [hunk 1]
```diff
@@ -2138,6 +2265,109 @@ function handleIdpSignUp(
...
+  return { mfaPendingCredential, mfaInfo: user.mfaInfo.map(redactMfaInfo) };
+}
+
+function redactMfaInfo(mfaInfo: MfaEnrollment): MfaEnrollment {
+  return {
+    displayName: mfaInfo.displayName,
+    enrolledAt: mfaInfo.enrolledAt,
+    mfaEnrollmentId: mfaInfo.mfaEnrollmentId,
+    phoneInfo: mfaInfo.unobfuscatedPhoneInfo
+      ? obfuscatePhoneNumber(mfaInfo.unobfuscatedPhoneInfo)
```
Comment (lisajian) [inline src/emulator/auth/operations.ts:2306]: Clarification question: I see that there's some [branching logic](https://source.corp.google.com/piper///depot/google3/java/com/google/identity/boq/idaas/mfa/MfaUtils.java;l=120-129;rcl=395089263) between unobfuscated vs obfuscated phone numbers - what was the reasoning behind defaulting to obfuscated numbers?

File: src/emulator/auth/operations.ts [hunk 2]
```diff
@@ -1597,6 +1594,136 @@ function listVerificationCodesInProject(
...
+): Schemas["GoogleCloudIdentitytoolkitV2StartMfaEnrollmentResponse"] {
+  assert(reqBody.idToken, "MISSING_ID_TOKEN");
+  assert(reqBody.phoneEnrollmentInfo, "((Missing phoneEnrollmentInfo.))");
+  // recaptchaToken, safetyNetToken, iosReceipt, and iosSecret are intentionally
+  // ignored because the emulator doesn't implement anti-abuse features.
+  // autoRetrievalInfo is ignored because SMS will not actually be sent.
+
+  const phoneNumber = reqBody.phoneEnrollmentInfo.phoneNumber;
+  assert(phoneNumber, "((Missing phoneNumber.))");
+  assert(isValidPhoneNumber(phoneNumber), "((Invalid phone number.)");
```
Comment (lisajian) [inline src/emulator/auth/operations.ts:1609]: Nit: missing closing `)` in assertion message
```suggestion
  assert(isValidPhoneNumber(phoneNumber), "((Invalid phone number.))");
```

File: src/emulator/auth/operations.ts [hunk 3]
```diff
@@ -1597,6 +1594,136 @@ function listVerificationCodesInProject(
...
+  assert(reqBody.mfaPendingCredential, "((Missing mfaPendingCredential.))");
+  assert(reqBody.mfaEnrollmentId, "((Missing mfaEnrollmentId.))");
+  const { user } = parsePendingCredential(state, reqBody.mfaPendingCredential);
+
+  const enrollment = user.mfaInfo?.find(
+    (factor) => factor.mfaEnrollmentId === reqBody.mfaEnrollmentId
+  );
+  assert(enrollment, "((mfaEnrollmentId not found.))");
+  const phoneNumber = enrollment.unobfuscatedPhoneInfo;
+  assert(phoneNumber, "((not a phone-based MFA enrollment.))");
```
Comment (lisajian) [inline src/emulator/auth/operations.ts:1685]: Nit on capitalization
```suggestion
  assert(phoneNumber, "((Not a phone-based MFA enrollment.))");
```

**Review by lisajian — COMMENTED** [review]
Would you mind double checking the following operations? They have mentions of [`MfaEnrollment`](https://source.corp.google.com/piper///depot/google3/google/cloud/identitytoolkit/v1/user_info.proto;l=168;bpv=1;bpt=1;rcl=395156284) in the response or request but it doesn't seem like there's a change present for them

* `resetPassword` - [response w/ `MfaEnrollment`](https://source.corp.google.com/piper///depot/google3/google/cloud/identitytoolkit/v1/account_management_service.proto;l=925;bpv=0;bpt=1;rcl=395156284)
* `setAccountInfo` - [request w/ `MfaEnrollment`](https://source.corp.google.com/piper///depot/google3/google/cloud/identitytoolkit/v1/account_management_service.proto;l=1090;bpv=0;bpt=1;rcl=395156284)

File: src/emulator/auth/operations.ts [hunk 4]
```diff
@@ -2138,6 +2265,109 @@ function handleIdpSignUp(
   };
 }
 
+type MfaEnrollment = Schemas["GoogleCloudIdentitytoolkitV1MfaEnrollment"];
+
+interface MfaPendingCredential {
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply src/emulator/auth/operations.ts:2342]: `signInProvider` is captured. Adding `projectId` for a later check. We probably don't need the other fields for now.

File: src/emulator/auth/operations.ts [hunk 5]
```diff
@@ -2138,6 +2265,109 @@ function handleIdpSignUp(
...
+function parsePendingCredential(
+  state: ProjectState,
+  pendingCredential: string
+): {
+  user: UserInfo;
+  signInProvider: string;
+} {
+  let pendingCredentialPayload: MfaPendingCredential;
+  try {
+    const json = Buffer.from(pendingCredential, "base64").toString("utf8");
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply src/emulator/auth/operations.ts:2428]: Ack but it's in general good practice to be explicit about encodings.

File: src/emulator/auth/operations.ts (hunk 1 above)
Comment (yuchenshi) [reply src/emulator/auth/operations.ts:2380]: From public docs: Normally this will show the phone number associated with this enrollment. In some situations, such as after a first factor sign in, it will only show the obfuscated version of the associated phone number.

File: src/emulator/auth/operations.ts [hunk 6]
```diff
@@ -2138,6 +2337,115 @@ function handleIdpSignUp(
...
+    phoneInfo: mfaInfo.unobfuscatedPhoneInfo
+      ? obfuscatePhoneNumber(mfaInfo.unobfuscatedPhoneInfo)
+      : undefined,
+  };
+}
+
+// Create an obfuscated version of a phone number, where all but the last
+// four digits are replaced by the character "*".
+function obfuscatePhoneNumber(phoneNumber: string): string {
+  let digitsLeft = 0;
```
Comment (sam-gc) [inline src/emulator/auth/operations.ts:2388]: I think this function could be simpler. For example this should also accomplish the same thing:
```ts
function obfuscatePhoneNumber(phoneNumber: string) {
  const split = phoneNumber.split('');
  let updated = 0;
  for (let i = split.length - 1; i >= 0; i--) {
    if (/[0-9]/.test(split[i])) {
      updated++;
      split[i] = '*';
    }
    if (updated === 4) {
      break;
    }
  }
  return split.join('');
}
```

File: src/emulator/auth/operations.ts [hunk 7]
```diff
@@ -1597,6 +1594,136 @@ function listVerificationCodesInProject(
...
+  });
+
+  const { idToken, refreshToken } = issueTokens(state, user, signInProvider);
+  return {
+    idToken,
+    refreshToken,
+  };
+}
+
+function mfaSignInStart(
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply src/emulator/auth/operations.ts:1760]: Added comment to clarify timeout is intentionally unimplemented

File: src/emulator/auth/operations.ts (hunk 6 above)
Comment (yuchenshi) [reply src/emulator/auth/operations.ts:2459]: Great idea, but your impl does the opposite -- fixed in PR

File: src/emulator/auth/state.ts [hunk 8]
```diff
@@ -375,11 +375,17 @@ export class ProjectState {
   createRefreshTokenFor(
     userInfo: UserInfo,
     provider: string,
-    extraClaims: Record<string, unknown> = {}
+    {
+      extraClaims = {},
+      secondFactor,
+    }: {
+      extraClaims?: Record<string, unknown>;
+      secondFactor?: { identifier: string; provider: string };
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply src/emulator/auth/state.ts]: Done

**Comment by yuchenshi** [comment]
> Would you mind double checking the following operations? They have mentions of [`MfaEnrollment`](http://go/firebase-tools-3732-mfaenrollment) in the response or request but it doesn't seem like there's a change present for them
> 
> * `resetPassword` - [response w/ `MfaEnrollment`](http://go/firebase-tools-3732-resetpassword)

I can look into this -- seems like an edge case though

> * `setAccountInfo` - [request w/ `MfaEnrollment`](http://go/firebase-tools-3732-setaccountinfo)

Already implemented in previous PRs.

**Review by sam-gc — APPROVED** [review]
Looks good to me :+1:
